### ISSUE TITLE
Is it possible to add rules to a parser at runtime?

### ISSUE THREAD

**Issue by jarble** [kach/nearley#485]
I want to write a parser that "learns" new grammar rules by modifying itself at runtime, but I don't know if Nearley can do this. An "adaptive grammar" rule might look like this:

`update_parser -> "add this rule: " expr`

When parsing `"add this rule: 'expr -> expr  "+" expr'"`, the parser would add 'expr -> expr + expr' as one of its grammar rules, so that it could parse strings that matched this pattern.

Is it currently possible to add rules to a parser at runtime?

**Comment by tjvr** [comment]
Yes! One of the nice benefits of an Earley parser is that there’s no pre-compilation step, so you can modify the grammar at runtime.

The caveat is that the behaviour may not be well-defined if you modify the grammar from inside a post-processor. Ideally you would split up your input somehow, and only modify the grammar in-between calls to `feed()`. Fiddly, but it can be done :)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by jarble** [comment]
@tjvr How is it possible to add rules to the grammar at runtime? I haven't found this feature in Nearley's documentation, but it would be a very useful feature to document.
